T30173 Rebase on 3.37.92

pwithnall opened the pull request:
This does most of the work for https://phabricator.endlessm.com/T30173, but isn’t actually the final rebase on 3.38.0 because that hasn’t been released yet.

Notable things to point out: demo mode has been dropped (as per https://phabricator.endlessm.com/T30173#840626), and the barcode has been dropped from the factory dialogue (as per https://phabricator.endlessm.com/T20124). Various smaller patches have been submitted upstream.

Packaging changes are in #318.

Here’s what happened to all the commits (oldest first), as a result of `git rebase -i upstream/3.37.91`:
 * **drop** `9378ab08 Allow to reorder pages from the vendor configuration file` — never used as per https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/48#note_677130
 * *pick* `a0576771 copypasta code: Use icon name instead of stock icon` — keep, can’t currently be upstreamed due to being blocked on https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/issues/68
 * *pick* `ab42158f copypasta code: Use GTask in UmRealmManager` — keep, can’t currently be upstreamed due to being blocked on https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/issues/68
 * *pick* `7f2bc934 Allow initial setup to switch metrics on and off` — keep, Endless-specific
 * **drop** `2fab2b48 summary: do not play welcome tour` — dropped in favour of https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/92, plus addition of `run_welcome_tour=false` to `debian/vendor.conf`
 * *pick* `615a973f Add product_serial.conf file` — keep, Endless-specific
 * *pick* `598105db account-local: limit full name to 80 characters` — keep, Endless-specific, and I don’t have the time to work through https://phabricator.endlessm.com/T25444 right now
 * *pick* `1ab4607e account-local: Hide username` — keep, Endless-specific for now
 * *pick* `e1d80c6d account-local: create a shared user account` — keep, Endless-specific for now
 * *pick* `23aff0a8 um-utils: consider "shared" username to be in use` — keep, Endless-specific for now
 * *pick* `40fa4f60 main,summary: skip initial setup entirely for shared account` — keep, Endless-specific for now
 * *pick* `c3915a4f account-local: username suggestions` — keep, but I have no idea what this is for and have not found time to investigate
 * *pick* `36376a2c password: Show the last character when entering the password` — keep, blocked on fixing https://phabricator.endlessm.com/T27358
 * *pick* `89cd30ea password: Add password reminder` — keep; upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/issues/56
 * *pick* `27c9370e Add gis_pkexec() helper function` — keep, Endless-specific
 * *pick* `de77e190 driver: track whether the current session is a live boot` — keep, Endless-specific
 * *pick* `efc5cf35 pages: Don't show some pages on live sessions` — keep, Endless-specific
 * **drop** `b519d6a9 assistant: add API to show/hide the forward button` — independently implemented upstream as 2ed5e1b7c5cfe74d87e76c01d12842421131da59
 * *pick* `f0f380d5 eos-test-mode: add eos-test-mode script` — keep, Endless-specific
 * *pick* `5f30a7f6 page-util: Add API to read the EOS image version` — keep, Endless-specific
 * *pick* `0d582ed6 endless-eula: Add endless eula page` — keep, Endless-specific
 * **drop** `1298d81d gnome-initial-setup: don't display privacy page` — dropped in favour of adding it to our `vendor.conf` file
 * *pick* `11fcd647 display: Add a "Display" page to the FBE` — keep; pulled in commit 00081c434c78205fedf0c0b36fa5d775cf7d496d from gnome-control-center; g-c-c files are otherwise unchanged upstream
 * *pick* `bbb1d968 keyboard: introduce detector` — keep, Endless-specific and I don’t have the time to upstream it right now
 * *pick* `9d17b6ee input-chooser: sort default layout before others` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `e1c40eed input-chooser: consider only language or country layouts non-extra` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `6c7bb941 input-chooser: Always show the selected layout` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `af846422 keyboard: add us/us+intl default layouts as fallback` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `75eaeb3c live-chooser: add the Live Chooser page` — keep, Endless-specific
 * *pick* `cc48bbc3 summary: show a different message on live sessions` — keep, Endless-specific
 * *pick* `1fe97a96 Allow pages to specify a GtkAccelGroup` — keep, Endless-specific
 * *pick* `dceddecc factory test mode: implement the factory dialog` — keep, Endless-specific; **note** significantly reworked to drop the barcode, as per https://phabricator.endlessm.com/T20124; and to use `g_get_os_info()`
 * *pick* `8685d26d language-chooser: make it more clear for the users` — keep; could be dropped in future if we updated `cc-language-chooser.c`
 * *pick* `a5aed058 timezone: set default 12h/24h from locale` — keep; could potentially be reworked and upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/47#note_583999, but I don’t have time right now
 * *pick* `ac330a78 Bring back the welcome message/logo` — keep, Endless-specific
 * **drop** `0f3de193 goa: reorder accounts` — rationale for this on https://phabricator.endlessm.com/T17486 no longer really applies: Facebook is not really important any more
 * *pick* `30d74b91 account-local: add toggle for passwordless accounts` — keep, Endless-specific
 * **fixup** `4f618ec6 fixup! account-local: add toggle for passwordless accounts` — squashed 
 * **fixup** `af40c727 fixup! account-local: add toggle for passwordless accounts` — squashed
 * **fixup** `3fbb910c fixup! account-local: add toggle for passwordless accounts` — squashed
 * *pick* `4836cb70 account,password: Hide password page for passwordless account` — keep, Endless-specific
 * *pick* `c2de4441 keyboard: manually pass locale to gkbd-keyboard-display` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `8979720e input-chooser: allow CcInputChooser to expand` — keep; could be dropped in future if we updated `cc-input-chooser.c`
 * *pick* `f7c6bf56 driver: Inhibit the idle state` — keep; could potentially be dropped if we worked on https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/issues/11, but I don’t have time for that right now
 * *pick* `88527b40 language: allow to specify initial languages from vendor.conf` — keep; could be upstreamed except it’s a modification to a file copied in from g-c-c so depends on https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/issues/68
 * **drop** `3f447d29 driver: Add "Store Demo"` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626
 * **drop** `988bbc6e copy-worker: Copy, as opposed to moving files` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626; attempted to upstream as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/94 but upstream decided not to
 * *pick* `a4715c3b language: make sure initial language is always at the top` — keep; could be dropped in future if we updated `cc-language-chooser.c`
 * *pick* `be2ea651 driver: detect whether live media is DVD` — keep, Endless-specific
 * *pick* `ec6b1e25 live-chooser: swap USB/DVD assets as needed` — keep, Endless-specific
 * **drop** `41963b9b endless-eula: Create, but don't show, EULA in store demo mode` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626
 * **drop** `4744029e language: Change the title to "Welcome to Store Demo" in store demo mode` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626
 * **drop** `83458d2c language: add a button to enter store demo mode on first page` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626
 * *pick* `b2c87d35 Don't run gnome-initial-setup for existing users` — keep; could probably be upstreamed as vendor config, but I have enough vendor config changes pending upstream for now
 * **drop** `238a6810 demo: Reset tracking-id and report a metric when we enable metrics` — dropping demo mode as per https://phabricator.endlessm.com/T30173#840626
 * *pick* `b2fd683c passwordPage: make latin kbd layout (en) available for password entries` — keep, but currently won’t work until https://phabricator.endlessm.com/T30649 is fixed
 * **drop** `4f523c79 Add powerMenu to initial-setup shell mode` — reverted by commit 1041f40f
 * *pick* `eaa6aa87 Launch eos-installer on eosinstaller images` — keep, Endless-specific
 * *pick* `792e0096 language: check {unattended,install}.ini on eosinstaller images` — keep, Endless-specific
 * *pick* `bf7a157a page-util: adopt logic to launch eos-installer` — keep, Endless-specific
 * *pick* `521dc3c0 password: Reset the password page whenever the password page gets hidden` — keep, Endless-specific until “account-local: add toggle for passwordless accounts” is upstreamed
 * *pick* `6d3f4db2 network: Update the description for WiFi` — keep; upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/95
 * *pick* `2d932003 site: add page to track site-specific details for deployments` — keep, Endless-specific; **note**: modified to drop demo mode support
 * *pick* `03d81387 account: Check configuration to hide password switch if needed` — keep, Endless-specific until “account-local: add toggle for passwordless accounts” is upstreamed
 * *pick* `708ddf57 timezone: Hide when automatically detected` — keep; could be upstreamed but not today
 * **drop** `256b0c09 account: Use control-center setting to override faces` — upstreamed as 4c020713141ef6745f02fe7421daf045fe5823e0
 * **drop** `c6735cf6 fixup! account: Use control-center setting to override faces` — upstreamed as 4c020713141ef6745f02fe7421daf045fe5823e0
 * *pick* `c8d81d89 driver: Make height 768 the threshold for "small screens"` — keep; upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/96
 * *pick* `c0ffe73a Change generated avatar for empty name` — keep, Endless-specific
 * *pick* `080f24f0 account: Add 'suggested-action' style to photo button` — keep; upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/97
 * *pick* `b513739f driver: Add resource base path` — keep, pending upstreaming at https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/43
 * *pick* `a566f521 password: Add eye icons to toggle password visibility` — keep, pending upstreaming at https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/43
 * **fixup** `42d8201d fixup! password: Add eye icons to toggle password visibility` — squashed
 * *pick* `3b5a1148 Add darker border to generated avatar` — keep; upstreamed as https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/99
 * *pick* `ed975d15 Configuration for Transifex` — keep, Endless-specific
 * *pick* `5822fc07 Allow gnome-initial-setup user to run eos-write-location` — keep, Endless-specific
 * *pick* `dbdf291f driver: detect when in live mode with persistence` — keep, Endless-specific
 * *pick* `9d027625 summary: show a different message on persistent live sessions` — keep, Endless-specific
 * *pick* `e29f48d9 Show all pages in persistent live mode` — keep, Endless-specific
 * *pick* `a30b8604 live-chooser: don't create live user account when in persistent live mode` — keep, Endless-specific
 * **drop** `735f6826 Limit names to 80 bytes to prevent adduser() failure` — squashed into “account-local: limit full name to 80 characters”
 * *pick* `680d5986 account: Make admin account usable in case of errors` — keep, Endless-specific
 * **drop** `d288c872 driver: Don't crash when saving parental controls` — upstreamed as bd2cf341e800d7b86112d8dac31f8a59cc000fb4
 * **drop** `49a6b048 driver: Allow setting parent password twice` — upstreamed as bd2cf341e800d7b86112d8dac31f8a59cc000fb4
 * *pick* `67acc01e eos: Report metrics from parental controls page` — keep, Endless-specific
 * **fixup** `b8c78cbc fixup! eos: Report metrics from parental controls page` — squashed
 * **drop** `bb4f8269 Add translation support for several new languages` — translations will be re-done after rebase
 * **drop** `ae24e403 (Re)generate .po files` — translations will be re-done after rebase
 * **drop** `8293c71d Rebase downstream translations` — translations will be re-done after rebase
 * **drop** `166038c2 Pull translations from Transifex` — translations will be re-done after rebase
 * **drop** `c5d6d8eb New translations from Transifex` — translations will be re-done after rebase
 * **drop** `eb1876e7 New translations from Transifex` — translations will be re-done after rebase
 * **drop** `dc35ea48 New translations from Transifex` — translations will be re-done after rebase
 * **drop** `1041f40f Revert "Add powerMenu to initial-setup shell mode"` — reverts an earlier commit
 * *pick* `34ad7874 data: Enable panel extension in initial-setup session` — keep, Endless-specific

pwithnall commented:
bot, retest this please

pwithnall commented:
bot, retest this please

mwleeds commented:
> pick 9d17b6ee input-chooser: sort default layout before others — keep; could be dropped in future if we updated cc-input-chooser.c
> pick e1c40eed input-chooser: consider only language or country layouts non-extra — keep; could be dropped in future if we updated cc-input-chooser.c
> pick 6c7bb941 input-chooser: Always show the selected layout — keep; could be dropped in future if we updated cc-input-chooser.c
> pick af846422 keyboard: add us/us+intl default layouts as fallback — keep; could be dropped in future if we updated cc-input-chooser.c

Might be worth adding a note to one or all of the commit messages with a bit more specificity than this (I'm not sure what you mean). And similarly for the other notes that say "if we updated <filename>", not sure what you mean.

> pick dceddecc factory test mode: implement the factory dialog — keep, Endless-specific; note significantly reworked to drop the barcode, as per https://phabricator.endlessm.com/T20124; and to use g_get_os_info()

Could make use of `g_auto`/`g_autofree` in `gis_page_util_show_factory_dialog()`.

> drop 0f3de193 goa: reorder accounts — rationale for this on https://phabricator.endlessm.com/T17486 no longer really applies: Facebook is not really important any more

This commit doesn't seem to have been dropped.

>pick 30d74b91 account-local: add toggle for passwordless accounts — keep, Endless-specific

The docs for `GtkCheckButton`/`GtkToggleButton` seem to say you should connect to the `toggled` signal but I can see how `notify::active` would work, so perhaps this is fine.

>pick 4836cb70 account,password: Hide password page for passwordless account — keep, Endless-specific

Shouldn't the `if (priv->parent_mode)` part of this check be upstream and independent of this commit?

>pick c2de4441 keyboard: manually pass locale to gkbd-keyboard-display — keep; could be dropped in future if we updated cc-input-chooser.c

Could get rid of some `goto` statements if you want

>pick b513739f driver: Add resource base path — keep, pending upstreaming at https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/43

You could add that to the commit message.

>pick 680d5986 account: Make admin account usable in case of errors — keep, Endless-specific

This is already upstream and the commit you have just adds the same lines that are already a few lines above.

(Sorry for reviewing code here rather than in the "Files Changed" part of the UI, but I am afraid to do it there because GitHub loses such comments if a force push happens.)

pwithnall commented:
I rebased again on `3.37.92`.

> > pick [9d17b6e](https://github.com/endlessm/gnome-initial-setup/commit/9d17b6ee107f0c9663ee7955c625b2fc673fb5ed) input-chooser: sort default layout before others — keep; could be dropped in future if we updated cc-input-chooser.c
> > pick [e1c40ee](https://github.com/endlessm/gnome-initial-setup/commit/e1c40eed201b1f6c1b3e44447496a5f47ad0459c) input-chooser: consider only language or country layouts non-extra — keep; could be dropped in future if we updated cc-input-chooser.c
> > pick [6c7bb94](https://github.com/endlessm/gnome-initial-setup/commit/6c7bb941342ed9e8de773213cbd53de0664baa10) input-chooser: Always show the selected layout — keep; could be dropped in future if we updated cc-input-chooser.c
> > pick [af84642](https://github.com/endlessm/gnome-initial-setup/commit/af846422b695b3da6a0e1f5a9392ff3884db637f) keyboard: add us/us+intl default layouts as fallback — keep; could be dropped in future if we updated cc-input-chooser.c
> 
> Might be worth adding a note to one or all of the commit messages with a bit more specificity than this (I'm not sure what you mean). And similarly for the other notes that say "if we updated ", not sure what you mean.

`cc-language-chooser` and `cc-input-chooser` are copy/paste code from gnome-control-center which has diverged over a long time. It’s a long-standing issue with gnome-initial-setup (upstream and downstream) that this code should be shared. Nobody has found time to fix it all up.

> > pick [dceddec](https://github.com/endlessm/gnome-initial-setup/commit/dceddecc984118610dbab2ffbf0978ec88b2a6bf) factory test mode: implement the factory dialog — keep, Endless-specific; note significantly reworked to drop the barcode, as per https://phabricator.endlessm.com/T20124; and to use g_get_os_info()
> 
> Could make use of `g_auto`/`g_autofree` in `gis_page_util_show_factory_dialog()`.

Well outside the scope of a rebase, but I’ve made a few improvements here.

> > drop [0f3de19](https://github.com/endlessm/gnome-initial-setup/commit/0f3de1935a5025878f2b9d704ddca3e0d0403a46) goa: reorder accounts — rationale for this on https://phabricator.endlessm.com/T17486 no longer really applies: Facebook is not really important any more
> 
> This commit doesn't seem to have been dropped.

Oops, dropped now.

> > pick [30d74b9](https://github.com/endlessm/gnome-initial-setup/commit/30d74b91fd6fd40397a2f351097117843bec69c6) account-local: add toggle for passwordless accounts — keep, Endless-specific
> 
> The docs for `GtkCheckButton`/`GtkToggleButton` seem to say you should connect to the `toggled` signal but I can see how `notify::active` would work, so perhaps this is fine.

I think they are equivalent, and the patch has worked in practice so I only see risk in changing it.

> > pick [4836cb7](https://github.com/endlessm/gnome-initial-setup/commit/4836cb700a19301b607cdd0371b3191d4a89482e) account,password: Hide password page for passwordless account — keep, Endless-specific
> 
> Shouldn't the `if (priv->parent_mode)` part of this check be upstream and independent of this commit?

I don’t think the situation `priv->parent_mode && !priv->username` can arise upstream — aiui the `!priv->username` code path is only taken when doing enterprise setups where there’s no local user being created, and parental controls aren’t supported on that code path at all (so you’ll never have `priv->parent_mode == TRUE`).

It can happen in this downstream patch, though, since that function has been changed to also handle property updates for passwordless accounts, which *can* be used in conjunction with parental controls.

> > pick [c2de444](https://github.com/endlessm/gnome-initial-setup/commit/c2de4441cc1e6f18a3b8aaf32465de49e40cf128) keyboard: manually pass locale to gkbd-keyboard-display — keep; could be dropped in future if we updated cc-input-chooser.c
> 
> Could get rid of some `goto` statements if you want

I’d rather not touch+diverge the `cc-input-chooser` code further from g-c-c.

> > pick [b513739](https://github.com/endlessm/gnome-initial-setup/commit/b513739f6ae3b63be57ee1753a6cbe3a18ccdf4a) driver: Add resource base path — keep, pending upstreaming at https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/43
> 
> You could add that to the commit message.

Good shout, done.

> > pick [680d598](https://github.com/endlessm/gnome-initial-setup/commit/680d5986ba80ebf898c0cab0b7130388b7cb6a6a) account: Make admin account usable in case of errors — keep, Endless-specific
> 
> This is already upstream and the commit you have just adds the same lines that are already a few lines above.

Huh, interesting that `git rebase` didn’t pick that up as a conflict. Dropped.

> (Sorry for reviewing code here rather than in the "Files Changed" part of the UI, but I am afraid to do it there because GitHub loses such comments if a force push happens.)

:+1: Thanks!

mwleeds commented on .gitlab-ci/fedora.Dockerfile:
```diff
@@ -0,0 +1,59 @@
+    "pkgconfig(accountsservice)" \
+    "pkgconfig(cheese)" \
+    "pkgconfig(cheese-gtk)" \
+    "pkgconfig(flatpak)" \
```
Looks like this was added upstream as part of the parental controls feature (9cd679a268e5627fb5d9abdf3e5eca9812d9fce4), but then the build dependency on `libflatpak-dev` was not kept during upstreaming so I guess it's not needed and this isn't either?

wjt replied:
libflatpak-dev is a build dependency of malcontent, which is compiled later down this dockerfile.
